File: dialects/f5/query/lexer.py
# tcl-lsp — a language server and toolchain for Tcl
# Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: AGPL-3.0-or-later

"""Tokeniser for the query DSL.

The grammar is small enough that a hand-rolled scanner is shorter than
the regex it would replace.  Tokens carry their offset so parse errors
can underline the offending character.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto

from .errors import LexError


class TokenKind(Enum):
    DOT = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    LPAREN = auto()
    RPAREN = auto()
    LBRACE = auto()
    RBRACE = auto()
    COMMA = auto()
    COLON = auto()
    SEMICOLON = auto()
    PIPE = auto()  # |
    PIPE_EQ = auto()  # |=
    PLUS = auto()
    PLUS_EQ = auto()
    MINUS = auto()
    MINUS_EQ = auto()
    STAR = auto()
    SLASH = auto()
    EQ = auto()  # =
    EQEQ = auto()  # ==
    NEQ = auto()  # !=
    LT = auto()
    LE = auto()
    GT = auto()
    GE = auto()
    STRING = auto()
    NUMBER = auto()
    IDENT = auto()  # bareword (may contain '-' after the first char)
    AND = auto()
    OR = auto()
    NOT = auto()
    TRUE = auto()
    FALSE = auto()
    NULL = auto()
    AS = auto()  # `expr as $name | body` — jq's let-binding form
    IF = auto()  # `if COND then BODY [elif ...] [else ...] end`
    THEN = auto()
    ELIF = auto()
    ELSE = auto()
    END = auto()
    DOLLAR_IDENT = auto()  # $name — variable reference to a named source
    QUESTION = auto()  # `?` — optional-marker suffix on a path step
    EOF = auto()


@dataclass(frozen=True, slots=True)
class Token:
    kind: TokenKind
    text: str
    offset: int

    # Convenience for string/number literals: the parsed value.
    value: object = None


_KEYWORDS = {
    "and": TokenKind.AND,
    "or": TokenKind.OR,
    "not": TokenKind.NOT,
    "true": TokenKind.TRUE,
    "false": TokenKind.FALSE,
    "null": TokenKind.NULL,
    "as": TokenKind.AS,
    "if": TokenKind.IF,
    "then": TokenKind.THEN,
    "elif": TokenKind.ELIF,
    "else": TokenKind.ELSE,
    "end": TokenKind.END,
}


def _is_ident_start(ch: str) -> bool:
    return ch.isalpha() or ch == "_"


def _is_ident_cont(ch: str) -> bool:
    # Hyphens are valid mid-identifier so that TMSH-spelt names like
    # ``data-group`` and ``source-address-translation`` lex as a single
    # token.  A leading hyphen is *not* an identifier — it is the minus
    # operator.
    return ch.isalnum() or ch in "_-"


def tokenise(source: str) -> list[Token]:
    out: list[Token] = []
    i = 0
    n = len(source)

    while i < n:
        ch = source[i]

        if ch.isspace():
            i += 1
            continue

        # Comments: ``# ...`` to end of line.  Useful for multi-line
        # queries loaded with ``-f``.
        if ch == "#":
            while i < n and source[i] != "\n":
                i += 1
            continue

        start = i

        # Single-character punctuation that has no two-char form.
        simple = {
            ".": TokenKind.DOT,
            "[": TokenKind.LBRACKET,
            "]": TokenKind.RBRACKET,
            "(": TokenKind.LPAREN,
            ")": TokenKind.RPAREN,
            "{": TokenKind.LBRACE,
            "}": TokenKind.RBRACE,
            ",": TokenKind.COMMA,
            ":": TokenKind.COLON,
            ";": TokenKind.SEMICOLON,
            "*": TokenKind.STAR,
            "/": TokenKind.SLASH,
            "?": TokenKind.QUESTION,
        }
        if ch in simple:
            out.append(Token(simple[ch], ch, start))
            i += 1
            continue

        # Two-char-or-one operators.
        if ch == "|":
            if i + 1 < n and source[i + 1] == "=":
                out.append(Token(TokenKind.PIPE_EQ, "|=", start))
                i += 2
            else:
                out.append(Token(TokenKind.PIPE, "|", start))
                i += 1
            continue

        if ch == "+":
            if i + 1 < n and source[i + 1] == "=":
                out.append(Token(TokenKind.PLUS_EQ, "+=", start))
                i += 2
            else:
                out.append(Token(TokenKind.PLUS, "+", start))
                i += 1
            continue

        if ch == "-":
            if i + 1 < n and source[i + 1] == "=":
                out.append(Token(TokenKind.MINUS_EQ, "-=", start))
                i += 2
            else:
                out.append(Token(TokenKind.MINUS, "-", start))
                i += 1
            continue

        if ch == "=":
            if i + 1 < n and source[i + 1] == "=":
                out.append(Token(TokenKind.EQEQ, "==", start))
                i += 2
            else:
                out.append(Token(TokenKind.EQ, "=", start))
                i += 1
            continue

        if ch == "!":
            if i + 1 < n and source[i + 1] == "=":
                out.append(Token(TokenKind.NEQ, "!=", start))
                i += 2
                continue
            raise LexError("unexpected '!' (did you mean '!='?)", start)

        if ch == "<":
            if i + 1 < n and source[i + 1] == "=":
                out.append(Token(TokenKind.LE, "<=", start))
                i += 2
            else:
                out.append(Token(TokenKind.LT, "<", start))
                i += 1
            continue

        if ch == ">":
            if i + 1 < n and source[i + 1] == "=":
                out.append(Token(TokenKind.GE, ">=", start))
                i += 2
            else:
                out.append(Token(TokenKind.GT, ">", start))
                i += 1
            continue

        # Variable references: ``$name``.  Used to address one specific
        # source when more than one config was loaded — e.g.
        # ``$gtm.gtm.wideip[]`` reads from the source bound under
        # ``gtm``.  The name follows the same identifier rules as a
        # bareword (hyphens allowed mid-name, underscores anywhere).
        if ch == "$":
            j = i + 1
            if j >= n or not _is_ident_start(source[j]):
                raise LexError(
                    "expected an identifier after '$' (e.g. $ltm)",
                    start,
                )
            j += 1
            while j < n and _is_ident_cont(source[j]):
                j += 1
            name = source[i + 1 : j]
            out.append(Token(TokenKind.DOLLAR_IDENT, source[i:j], start, value=name))
            i = j
            continue

        # Strings: double-quoted, with the usual backslash escapes.
        # A leading ``~`` is *not* a token kind of its own — the parser
        # checks for it when the string appears as the body of a
        # subscript ``[ ... ]``.  That keeps ``sub(.x, "~lit", "y")``
        # working: the literal happens to start with ``~``, but it is
        # not a regex subscript and is passed through unchanged.
        if ch == '"':
            j = i + 1
            buf: list[str] = []
            while j < n and source[j] != '"':
                if source[j] == "\\" and j + 1 < n:
                    esc = source[j + 1]
                    buf.append(
                        {
                            "n": "\n",
                            "t": "\t",
                            "r": "\r",
                            '"': '"',
                            "\\": "\\",
                        }.get(esc, esc)
                    )
                    j += 2
                    continue
                buf.append(source[j])
                j += 1
            if j >= n:
                raise LexError("unterminated string literal", start)
            text = "".join(buf)
            j += 1  # consume closing quote
            out.append(Token(TokenKind.STRING, source[start:j], start, value=text))
            i = j
            continue

        # Numbers.  Integer or float (no scientific notation in v1).
        if ch.isdigit():
            j = i
            while j < n and source[j].isdigit():
                j += 1
            if j < n and source[j] == "." and j + 1 < n and source[j + 1].isdigit():
                j += 1
                while j < n and source[j].isdigit():
                    j += 1
                value: object = float(source[i:j])
            else:
                value = int(source[i:j])
            out.append(Token(TokenKind.NUMBER, source[i:j], start, value=value))
            i = j
            continue

        # Identifiers / keywords.
        if _is_ident_start(ch):
            j = i + 1
            while j < n and _is_ident_cont(source[j]):
                j += 1
            text = source[i:j]
            kw = _KEYWORDS.get(text)
            if kw is not None:
                value = {
                    TokenKind.TRUE: True,
                    TokenKind.FALSE: False,
                    TokenKind.NULL: None,
                }.get(kw)
                out.append(Token(kw, text, start, value=value))
            else:
                out.append(Token(TokenKind.IDENT, text, start, value=text))
            i = j
            continue

        raise LexError(f"unexpected character {ch!r}", start)

    out.append(Token(TokenKind.EOF, "", n))
    return out

File: rust/bigip/tcl-bigip-query/src/lexer.rs
// tcl-lsp — a language server and toolchain for Tcl
// Copyright (C) 2026 James Deucker (bitwisecook) <https://github.com/bitwisecook>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Tokeniser for the query DSL.
//!
//! The grammar is small enough that a hand-rolled scanner is shorter than the
//! regex it would replace. Tokens carry their offset so parse errors can
//! underline the offending character.
//!
//! Offsets are *code-point* indices (the source is scanned as a
//! `Vec<char>`), matching `source[i]` indexing exactly so a
//! non-ASCII string literal does not skew the reported positions.

use crate::ast::LitValue;
use crate::errors::QueryError;

/// The lexical category of a token.
///
/// Variant names match `TokenKind` enum members one-for-one so
/// the differential fixtures line up.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Dot,
    LBracket,
    RBracket,
    LParen,
    RParen,
    LBrace,
    RBrace,
    Comma,
    Colon,
    Semicolon,
    Pipe,
    PipeEq,
    Plus,
    PlusEq,
    Minus,
    MinusEq,
    Star,
    Slash,
    Eq,
    EqEq,
    Neq,
    Lt,
    Le,
    Gt,
    Ge,
    String,
    Number,
    Ident,
    And,
    Or,
    Not,
    True,
    False,
    Null,
    As,
    If,
    Then,
    Elif,
    Else,
    End,
    DollarIdent,
    Question,
    Eof,
}

impl TokenKind {
    /// The `TokenKind` member name, used by the differential test.
    #[must_use]
    pub fn name(self) -> &'static str {
        match self {
            TokenKind::Dot => "DOT",
            TokenKind::LBracket => "LBRACKET",
            TokenKind::RBracket => "RBRACKET",
            TokenKind::LParen => "LPAREN",
            TokenKind::RParen => "RPAREN",
            TokenKind::LBrace => "LBRACE",
            TokenKind::RBrace => "RBRACE",
            TokenKind::Comma => "COMMA",
            TokenKind::Colon => "COLON",
            TokenKind::Semicolon => "SEMICOLON",
            TokenKind::Pipe => "PIPE",
            TokenKind::PipeEq => "PIPE_EQ",
            TokenKind::Plus => "PLUS",
            TokenKind::PlusEq => "PLUS_EQ",
            TokenKind::Minus => "MINUS",
            TokenKind::MinusEq => "MINUS_EQ",
            TokenKind::Star => "STAR",
            TokenKind::Slash => "SLASH",
            TokenKind::Eq => "EQ",
            TokenKind::EqEq => "EQEQ",
            TokenKind::Neq => "NEQ",
            TokenKind::Lt => "LT",
            TokenKind::Le => "LE",
            TokenKind::Gt => "GT",
            TokenKind::Ge => "GE",
            TokenKind::String => "STRING",
            TokenKind::Number => "NUMBER",
            TokenKind::Ident => "IDENT",
            TokenKind::And => "AND",
            TokenKind::Or => "OR",
            TokenKind::Not => "NOT",
            TokenKind::True => "TRUE",
            TokenKind::False => "FALSE",
            TokenKind::Null => "NULL",
            TokenKind::As => "AS",
            TokenKind::If => "IF",
            TokenKind::Then => "THEN",
            TokenKind::Elif => "ELIF",
            TokenKind::Else => "ELSE",
            TokenKind::End => "END",
            TokenKind::DollarIdent => "DOLLAR_IDENT",
            TokenKind::Question => "QUESTION",
            TokenKind::Eof => "EOF",
        }
    }
}

/// A lexed token: its kind, source text, code-point offset, and — for
/// string / number / keyword / identifier tokens — the parsed value.
#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub text: String,
    pub offset: usize,
    /// The parsed value (`LitValue::Null` when the token carries none).
    pub value: LitValue,
}

fn keyword(text: &str) -> Option<TokenKind> {
    match text {
        "and" => Some(TokenKind::And),
        "or" => Some(TokenKind::Or),
        "not" => Some(TokenKind::Not),
        "true" => Some(TokenKind::True),
        "false" => Some(TokenKind::False),
        "null" => Some(TokenKind::Null),
        "as" => Some(TokenKind::As),
        "if" => Some(TokenKind::If),
        "then" => Some(TokenKind::Then),
        "elif" => Some(TokenKind::Elif),
        "else" => Some(TokenKind::Else),
        "end" => Some(TokenKind::End),
        _ => None,
    }
}

fn is_ident_start(ch: char) -> bool {
    ch.is_alphabetic() || ch == '_'
}

fn is_ident_cont(ch: char) -> bool {
    // Hyphens are valid mid-identifier so TMSH-spelt names like
    // `data-group` lex as a single token. A leading hyphen is *not* an
    // identifier — it is the minus operator.
    ch.is_alphanumeric() || ch == '_' || ch == '-'
}

fn simple_punct(ch: char) -> Option<TokenKind> {
    match ch {
        '.' => Some(TokenKind::Dot),
        '[' => Some(TokenKind::LBracket),
        ']' => Some(TokenKind::RBracket),
        '(' => Some(TokenKind::LParen),
        ')' => Some(TokenKind::RParen),
        '{' => Some(TokenKind::LBrace),
        '}' => Some(TokenKind::RBrace),
        ',' => Some(TokenKind::Comma),
        ':' => Some(TokenKind::Colon),
        ';' => Some(TokenKind::Semicolon),
        '*' => Some(TokenKind::Star),
        '/' => Some(TokenKind::Slash),
        '?' => Some(TokenKind::Question),
        _ => None,
    }
}

fn escape(esc: char) -> char {
    match esc {
        'n' => '\n',
        't' => '\t',
        'r' => '\r',
        '"' => '"',
        '\\' => '\\',
        other => other,
    }
}

/// Tokenise *source* into a token stream ending in an `Eof` token.
///
/// # Errors
///
/// Returns [`QueryError::Lex`] on an unterminated string, a stray `!` /
/// `$`, or any character the scanner does not understand.
#[allow(clippy::too_many_lines)]
pub fn tokenise(source: &str) -> Result<Vec<Token>, QueryError> {
    let chars: Vec<char> = source.chars().collect();
    let n = chars.len();
    let mut out: Vec<Token> = Vec::new();
    let mut i = 0usize;

    let slice = |a: usize, b: usize| -> String { chars[a..b].iter().collect() };

    while i < n {
        let ch = chars[i];

        if ch.is_whitespace() {
            i += 1;
            continue;
        }

        // Comments: `# ...` to end of line.
        if ch == '#' {
            while i < n && chars[i] != '\n' {
                i += 1;
            }
            continue;
        }

        let start = i;

        // Single-character punctuation that has no two-char form.
        if let Some(kind) = simple_punct(ch) {
            out.push(Token {
                kind,
                text: ch.to_string(),
                offset: start,
                value: LitValue::Null,
            });
            i += 1;
            continue;
        }

        // Two-char-or-one operators.
        if ch == '|' {
            if i + 1 < n && chars[i + 1] == '=' {
                out.push(Token {
                    kind: TokenKind::PipeEq,
                    text: "|=".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 2;
            } else {
                out.push(Token {
                    kind: TokenKind::Pipe,
                    text: "|".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 1;
            }
            continue;
        }

        if ch == '+' {
            if i + 1 < n && chars[i + 1] == '=' {
                out.push(Token {
                    kind: TokenKind::PlusEq,
                    text: "+=".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 2;
            } else {
                out.push(Token {
                    kind: TokenKind::Plus,
                    text: "+".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 1;
            }
            continue;
        }

        if ch == '-' {
            if i + 1 < n && chars[i + 1] == '=' {
                out.push(Token {
                    kind: TokenKind::MinusEq,
                    text: "-=".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 2;
            } else {
                out.push(Token {
                    kind: TokenKind::Minus,
                    text: "-".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 1;
            }
            continue;
        }

        if ch == '=' {
            if i + 1 < n && chars[i + 1] == '=' {
                out.push(Token {
                    kind: TokenKind::EqEq,
                    text: "==".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 2;
            } else {
                out.push(Token {
                    kind: TokenKind::Eq,
                    text: "=".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 1;
            }
            continue;
        }

        if ch == '!' {
            if i + 1 < n && chars[i + 1] == '=' {
                out.push(Token {
                    kind: TokenKind::Neq,
                    text: "!=".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 2;
                continue;
            }
            return Err(QueryError::Lex {
                message: "unexpected '!' (did you mean '!='?)".to_string(),
                offset: start,
            });
        }

        if ch == '<' {
            if i + 1 < n && chars[i + 1] == '=' {
                out.push(Token {
                    kind: TokenKind::Le,
                    text: "<=".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 2;
            } else {
                out.push(Token {
                    kind: TokenKind::Lt,
                    text: "<".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 1;
            }
            continue;
        }

        if ch == '>' {
            if i + 1 < n && chars[i + 1] == '=' {
                out.push(Token {
                    kind: TokenKind::Ge,
                    text: ">=".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 2;
            } else {
                out.push(Token {
                    kind: TokenKind::Gt,
                    text: ">".to_string(),
                    offset: start,
                    value: LitValue::Null,
                });
                i += 1;
            }
            continue;
        }

        // Variable references: `$name`.
        if ch == '$' {
            let mut j = i + 1;
            if j >= n || !is_ident_start(chars[j]) {
                return Err(QueryError::Lex {
                    message: "expected an identifier after '$' (e.g. $ltm)".to_string(),
                    offset: start,
                });
            }
            j += 1;
            while j < n && is_ident_cont(chars[j]) {
                j += 1;
            }
            let name = slice(i + 1, j);
            out.push(Token {
                kind: TokenKind::DollarIdent,
                text: slice(i, j),
                offset: start,
                value: LitValue::Str(name),
            });
            i = j;
            continue;
        }

        // Strings: double-quoted, with the usual backslash escapes.
        if ch == '"' {
            let mut j = i + 1;
            let mut buf = String::new();
            while j < n && chars[j] != '"' {
                if chars[j] == '\\' && j + 1 < n {
                    buf.push(escape(chars[j + 1]));
                    j += 2;
                    continue;
                }
                buf.push(chars[j]);
                j += 1;
            }
            if j >= n {
                return Err(QueryError::Lex {
                    message: "unterminated string literal".to_string(),
                    offset: start,
                });
            }
            j += 1; // consume closing quote
            out.push(Token {
                kind: TokenKind::String,
                text: slice(start, j),
                offset: start,
                value: LitValue::Str(buf),
            });
            i = j;
            continue;
        }

        // Numbers. Integer or float (no scientific notation in v1).
        if ch.is_ascii_digit() {
            let mut j = i;
            while j < n && chars[j].is_ascii_digit() {
                j += 1;
            }
            let value: LitValue =
                if j < n && chars[j] == '.' && j + 1 < n && chars[j + 1].is_ascii_digit() {
                    j += 1;
                    while j < n && chars[j].is_ascii_digit() {
                        j += 1;
                    }
                    let text = slice(i, j);
                    // A run of digits with a single dot is always a valid
                    // `f64` lexeme; out-of-range magnitudes saturate to
                    // ±inf rather than failing, so `parse` cannot error here.
                    // Guard it anyway so a future grammar change can never
                    // turn a malformed number into a panic.
                    match text.parse::<f64>() {
                        Ok(f) => LitValue::Float(f),
                        Err(_) => {
                            return Err(QueryError::Lex {
                                message: format!("invalid number literal {}", py_repr_str(&text)),
                                offset: start,
                            });
                        }
                    }
                } else {
                    let text = slice(i, j);
                    // The integer literal is untrusted user input: a digit run
                    // longer than `i64::MAX` (`99999999999999999999`) overflows
                    // the parse. Report a clean lex error instead of panicking
                    // — this DSL has no big-int model, so the value is simply
                    // out of range.
                    match text.parse::<i64>() {
                        Ok(int) => LitValue::Int(int),
                        Err(_) => {
                            return Err(QueryError::Lex {
                                message: format!(
                                    "integer literal {} is out of range for a 64-bit integer",
                                    py_repr_str(&text)
                                ),
                                offset: start,
                            });
                        }
                    }
                };
            out.push(Token {
                kind: TokenKind::Number,
                text: slice(i, j),
                offset: start,
                value,
            });
            i = j;
            continue;
        }

        // Identifiers / keywords.
        if is_ident_start(ch) {
            let mut j = i + 1;
            while j < n && is_ident_cont(chars[j]) {
                j += 1;
            }
            let text = slice(i, j);
            if let Some(kw) = keyword(&text) {
                let value = match kw {
                    TokenKind::True => LitValue::Bool(true),
                    TokenKind::False => LitValue::Bool(false),
                    _ => LitValue::Null,
                };
                out.push(Token {
                    kind: kw,
                    text,
                    offset: start,
                    value,
                });
            } else {
                out.push(Token {
                    kind: TokenKind::Ident,
                    text: text.clone(),
                    offset: start,
                    value: LitValue::Str(text),
                });
            }
            i = j;
            continue;
        }

        return Err(QueryError::Lex {
            message: format!("unexpected character {}", py_repr_str(&ch.to_string())),
            offset: start,
        });
    }

    out.push(Token {
        kind: TokenKind::Eof,
        text: String::new(),
        offset: n,
        value: LitValue::Null,
    });
    Ok(out)
}

/// Render *s* the way `repr()` would for a `str` — the spelling
/// the DSL's error messages embed via `{value!r}`. Single-quoted by
/// default, switching to double quotes only when the string contains a
/// single quote but no double quote (the single-quote choice), with
/// backslash / newline / tab / carriage-return escaping.
///
/// Shared by the lexer (`unexpected character {ch!r}`) and the parser
/// (`unexpected token {tok.text!r}`, the `as`-form diagnostics, …).
pub(crate) fn py_repr_str(s: &str) -> String {
    let has_single = s.contains('\'');
    let has_double = s.contains('"');
    let quote = if has_single && !has_double { '"' } else { '\'' };
    let mut out = String::new();
    out.push(quote);
    for ch in s.chars() {
        match ch {
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            c if c == quote => {
                out.push('\\');
                out.push(c);
            }
            c => out.push(c),
        }
    }
    out.push(quote);
    out
}
